src/main/java/Group15/Util/WorkoutAlgorithm.java

```diff
@@ -36,36 +33,39 @@ public static Workout createWorkoutFromExercises(
             workoutExercise.setRepsPerSet(5);
             int timePerSet = validExercise.timePerRep * 5;
 
-            while (timeLeftInMilli > 0 && workoutExercise.getSets() < 3)
-            {
-                timeLeftInMilli -= timePerSet;
+            while (initialTimeInMilli - workout.getWorkoutDuration() > 0 && workoutExercise.getSets() < MAX_SETS) {
                 workoutExercise.setSets(workoutExercise.getSets() + 1);
-                timeLeftInMilli -= breakBetweenSets;
+
+                int timeUsedThisSet = workout.getWorkoutDuration() + timePerSet;
```

Comment:
This is does not appear to be the time used this set but time used in total up to that point.
Plus you are calling getWorkoutDuration a few times here, which is not very efficient as the method has to iterate through the whole workout every time.